source/common/unified_memory_allocation/src/memory_provider_internal.h

```diff
@@ -0,0 +1,24 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ */
+
+#ifndef UMA_MEMORY_PROVIDER_INTERNAL_H
+#define UMA_MEMORY_PROVIDER_INTERNAL_H 1
+
+#include <uma/memory_provider.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void* umaMemoryProviderGetPriv(uma_memory_provider_handle_t hProvider);
```

Comment:
I don't like that I had to add this function, but I couldn't; come up with a better way - maybe you have some ideas?

The problem is that we need to call a memory tracking provider-specific function that is implemented in memory_tracker.cpp  (getUpstreamProvider). However, this function does not know anything about how `uma_memory_provider_handle_t` is implemented (it's hidden in memory_provider.c) so we need a way to get the internal pointer (tracker-specific).

I think this might be a problem for the user, as well (right now this function is internal but maybe we need to expose it?). What if he wants to call his provider-specific function? He cannot do that if he only has a `uma_memory_provider_handle_t`. There is no way for the user to get a pointer to back his provider class/structure.

Reply:
I had to spent some time to even understand why we need to get the priv pointer. I agree it's confusing, but I'm not sure there's an alternative - we lost type information about the provider when we create an uma handle. The best approach would be to allow providers to somehow identify themselves (some unique string or id), that then can be used at runtime to match with the correct type - e.g.,:
```
provider providers[] = {a, b, c};

for (p in providers) {
   if (p.implementation_id() == unix_provider.id()) {
          unix_provider_custom_func(p);
   }
}
```
But that's probably too verbose to be useful.

Anyway, I think we should have a get_priv() method. I'd also go a step further and change the provider (and heap) ops - https://github.com/oneapi-src/unified-runtime/blob/main/source/common/unified_memory_allocation/include/uma/memory_provider_ops.h#L21 - to use the handle and internally the providers can use get_priv to get their data. I find the current implementation confusing... We use the variable name 'provider' to mean both the handle and the provider priv data.

Reply:
I'm not sure about making provider/pool call get_priv: you'd need to link every pool/provider with uma_core. Right now, we only have a dependency on the uma_base.h header.

Also, passing `priv` to each provider/pool seems very similar to what UR is doing with adapters. Pehaps we should even rename the `priv` to `native` and have `uma_pool/provider_native_handel_t` instead of `void*`?